src/simbricks/init_manager.cc

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright 2022 Max Planck Institute for Software Systems, and
+ * National University of Singapore
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "simbricks/init_manager.hh"
+
+#include <poll.h>
+
+#include <csignal>
+
+namespace gem5 {
+namespace simbricks {
+namespace base {
+
+static void sigusr1_handler(int dummy)
+{
+    std::cout << "main_time = " << curTick() << std::endl;
+}
+
+InitManager::InitManager()
+{
+}
+
+InitManager&
+InitManager::get()
+{
+    static InitManager *im = nullptr;
+    if (!im) {
+        im = new InitManager();
+
+        // also register usr1 signal handler (once)
+        signal(SIGUSR1, sigusr1_handler);
+    }
+
+    return *im;
+}
+
+void
+InitManager::registerAdapter(Adapter &a)
+{
+    unconnected.insert(&a);
+}
+
+void
+InitManager::connected(Adapter &a)
+{
+    waitRx.insert(&a);
+    unconnected.erase(&a);
+
+    const unsigned max_handshake = 4069;
+    std::vector<uint8_t> handshake(max_handshake);
```

Comment:
Using an std::vector here to allocate the data seems very weird. How about the following instead?

```suggestion
    uint8_t handshake[max_handshake];
```

Reply:
Having kilobytes worth of up of data on the stack is bad practice. Byte vector seems like the c++ way of allocating dynamic byte buffers. Unless someone wants to fix work something clean.

Reply:
I just discussed this with @q713 and @marvin71 and my opinion is now also that we should leave it as is. I noticed that `max_handshake` is set to `4069` though. Is this intentional or should it actually be `4096`?

Reply:
Good catch. Fixed.